nemo_aligner/algorithms/rs.py

```diff
@@ -0,0 +1,480 @@
+            mask = create_mask(
+                values=torch.zeros([response_tokens.shape[0], response_tokens.shape[1] - 1]),
+                prompt_lengths=prompt_lengths,
+                response_lengths=response_lengths,
+            )
+
+            # collect everything we need to train actor
+            rs_rollout_data["mask"].extend(post_process_tensor(mask))
+            rs_rollout_data["response_tokens"].extend(post_process_tensor(response_tokens))
+            rs_rollout_data["prompt_tokens"].extend(post_process_tensor(prompt_tokens))
+
+        # average across the samples for the non global metrics
+        rs_rollout_metrics = {k: v / num_samples for k, v in rs_rollout_metrics.items()}
+
+        for k in rs_rollout_data:
+            rollout_batch_seq_length = self.rollout_batch_seq_length
+            pad_value = self.model.tokenizer.eos_id
+
+            # all other tensors in the rollout batch
+            # will be B x S -1 (because we don't predict anything for the last token)
+
+            rs_rollout_data[k] = pad_tensors_to_max_global_seq_len(
+                rs_rollout_data[k],
+                pad_value=pad_value,
+                group=parallel_state.get_data_parallel_group(),
+                sequence_length_to_pad_to=rollout_batch_seq_length,
+            )
+
+        return rs_rollout_data, cpu_dict(rs_rollout_metrics)
+
+    def _run_inference(self, dataloader_iter, num_microbatches, is_validation):
+        """this function is run per DP so the metrics need to be computed globally
+        """
+        rollout_batches = []
+        if not is_validation:
+            full_batches = []  # compute metrics over all batches, not just the selected ones
+            for _, inference_batch in zip(range(num_microbatches), dataloader_iter):
+
+                current_batch = None
+                prompt_tokens, response_tokens, response_lengths, prompt_lengths, rewards = (
+                    [],
+                    [],
+                    [],
+                    [],
+                    [],
+                )
+                for _ in range(self.num_rollouts_per_prompt):
+                    rollout_batch = self.model.infer(inference_batch)
+                    reward = self.rm.infer_rm(rollout_batch).result().detach()
+
+                    prompt_tokens.append(inference_batch["text"])
+                    response_tokens.append(rollout_batch["response_tokens"])
+                    response_lengths.append(rollout_batch["response_lengths"])
+                    prompt_lengths.append(rollout_batch["prompt_lengths"])
+                    rewards.append(reward)
+
+                all_rollouts = {}
+                all_rollouts["response_tokens"] = torch.concatenate(
+                    pad_list(response_tokens, pad_value=self.model.tokenizer.eos_id)
+                )
+                all_rollouts["prompt_tokens"] = torch.concatenate(
+                    pad_list(prompt_tokens, pad_value=self.model.tokenizer.eos_id)
```

Comment:
Why do you need padding for `prompt_tokens`? In your original implementation you were just doing
```python
current_batch["prompt_tokens"] = torch.concatenate(
    [current_batch["prompt_tokens"], inference_batch["text"]], dim=0
)
```

Reply:
You are right, it is not needed. I have removed it.